### PR TITLE
add time in output lines

### PR DESCRIPTION
**Pull request by SYSHIL**
# Important

**Please provide the requested parameters below to help us review your pull request.**

**Pull requests without a body or explanation will be rejected.**

## Issue ticket number and link

github.com/pouriyajamshidi/tcping/issues/215

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [ ] If it is a core feature, I have added tests.
- [x] I have run `make check` and there are no failures.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,4 @@ tcping.exe
 *.code-workspace
 patches/
 *.db
```
Comment (SYSHIL) [inline .gitignore:10]: ignore .tool-versions is useful for folks like me that use asdf :)
I hope this is okay?

**Comment by SYSHIL** [comment]
Output with the new changes looks something like this. @pouriyajamshidi could you please review? Thanks in advance :)


```
ilhan@syeds-MacBook-Pro tcping % go run . www.example.com 80 
2024-08-03 00:31:17 TCPinging www.example.com on port 80
2024-08-03 00:31:18 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=1 time=303.326 ms
2024-08-03 00:31:19 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=2 time=263.166 ms
2024-08-03 00:31:20 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=3 time=350.889 ms
2024-08-03 00:31:21 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=4 time=272.626 ms
2024-08-03 00:31:22 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=5 time=296.352 ms
2024-08-03 00:31:23 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=6 time=256.992 ms
2024-08-03 00:31:24 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=7 time=300.664 ms
2024-08-03 00:31:24 No reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=1
2024-08-03 00:31:25 No reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=2
2024-08-03 00:31:26 No reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=3
2024-08-03 00:31:27 No reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=4
2024-08-03 00:31:28 No reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=5
2024-08-03 00:31:29 No reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=6
2024-08-03 00:31:31 No response received for 6 seconds
2024-08-03 00:31:31 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=1 time=929.736 ms
2024-08-03 00:31:32 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=2 time=451.907 ms
2024-08-03 00:31:33 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=3 time=480.420 ms
2024-08-03 00:31:34 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=4 time=377.191 ms

2024-08-03 00:31:35 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=5 time=298.571 ms

--- www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) TCPing statistics ---
18 probes transmitted on port 80 | 12 received, 33.33% packet loss
successful probes:   12
unsuccessful probes: 6
last successful probe:   2024-08-03 00:31:34
last unsuccessful probe: 2024-08-03 00:31:29
total uptime:   12 seconds
total downtime: 6 seconds
longest consecutive uptime:   7 seconds from 2024-08-03 00:31:17 to 2024-08-03 00:31:24
longest consecutive downtime: 6 seconds from 2024-08-03 00:31:24 to 2024-08-03 00:31:30
retried to resolve hostname 0 times
rtt min/avg/max: 256.992/381.820/929.736 ms
--------------------------------------
TCPing started at: 2024-08-03 00:31:17
duration (HH:MM:SS): 00:00:18

2024-08-03 00:31:36 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=6 time=240.240 ms
2024-08-03 00:31:37 Reply from www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) on port 80 TCP_conn=7 time=250.638 ms
^C
--- www.example.com (2606:2800:21f:cb07:6820:80da:af6b:8b2c) TCPing statistics ---
20 probes transmitted on port 80 | 14 received, 30.00% packet loss
successful probes:   14
unsuccessful probes: 6
last successful probe:   2024-08-03 00:31:36
last unsuccessful probe: 2024-08-03 00:31:29
total uptime:   14 seconds
total downtime: 6 seconds
longest consecutive uptime:   7 seconds from 2024-08-03 00:31:17 to 2024-08-03 00:31:24
longest consecutive downtime: 6 seconds from 2024-08-03 00:31:24 to 2024-08-03 00:31:30
retried to resolve hostname 0 times
rtt min/avg/max: 240.240/362.337/929.736 ms
--------------------------------------
TCPing started at: 2024-08-03 00:31:17
TCPing ended at:   2024-08-03 00:31:37
duration (HH:MM:SS): 00:00:20
```

**Comment by pouriyajamshidi** [comment]
Hi @SYSHIL,

I had a very quick look at your PR and the first thing that jumped at me is that the default behavior has changed.

This is not the intention.

We try to stay relatively close to what Linux's ping utility does and as the issue was describing, timestamps are only activated using the `-D` flag.

Let's address the rest of the open topics after that.

**Comment by SYSHIL** [comment]
Understood, Thank you @pouriyajamshidi I'll look into this further

**Comment by SYSHIL** [comment]
Hey @pouriyajamshidi I added a boolean field in the plane printer to ensure the default behaviour doesn't get affected and the timestamp is only added to our logs when a -D flag is passed.
here is the sample usage, could you please check?

```
2024-08-05 19:21:49 TCPinging www.google.com on port 80
2024-08-05 19:21:49 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=1
2024-08-05 19:21:50 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=2
2024-08-05 19:21:51 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=3
2024-08-05 19:21:52 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=4
2024-08-05 19:21:53 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=5
2024-08-05 19:21:54 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=6
2024-08-05 19:21:55 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=7
2024-08-05 19:21:56 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=8
2024-08-05 19:21:57 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=9
2024-08-05 19:21:58 No reply from www.google.com (142.250.195.196) on port 80 TCP_conn=1
2024-08-05 19:21:59 No reply from www.google.com (142.250.195.196) on port 80 TCP_conn=2
2024-08-05 19:22:00 No reply from www.google.com (142.250.195.196) on port 80 TCP_conn=3
2024-08-05 19:22:01 No reply from www.google.com (142.250.195.196) on port 80 TCP_conn=4
2024-08-05 19:22:02 No reply from www.google.com (142.250.195.196) on port 80 TCP_conn=5
2024-08-05 19:22:03 No response received for 5 seconds
2024-08-05 19:22:03 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=1
2024-08-05 19:22:04 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=2
2024-08-05 19:22:05 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=3
2024-08-05 19:22:06 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=4
2024-08-05 19:22:07 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=5
```

**Comment by pouriyajamshidi** [comment]
> Hey @pouriyajamshidi I added a boolean field in the plane printer to ensure the default behaviour doesn't get affected and the timestamp is only added to our logs when a -D flag is passed. here is the sample usage, could you please check? in case its okay i can update our READMEs too
> 
> ```
> 2024-08-05 19:21:49 TCPinging www.google.com on port 80
> 2024-08-05 19:21:49 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=1
> 2024-08-05 19:21:50 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=2
> 2024-08-05 19:21:51 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=3
> 2024-08-05 19:21:52 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=4
> 2024-08-05 19:21:53 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=5
> 2024-08-05 19:21:54 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=6
> 2024-08-05 19:21:55 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=7
> 2024-08-05 19:21:56 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=8
> 2024-08-05 19:21:57 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=9
> 2024-08-05 19:21:58 No reply from www.google.com (142.250.195.196) on port 80 TCP_conn=1
> 2024-08-05 19:21:59 No reply from www.google.com (142.250.195.196) on port 80 TCP_conn=2
> 2024-08-05 19:22:00 No reply from www.google.com (142.250.195.196) on port 80 TCP_conn=3
> 2024-08-05 19:22:01 No reply from www.google.com (142.250.195.196) on port 80 TCP_conn=4
> 2024-08-05 19:22:02 No reply from www.google.com (142.250.195.196) on port 80 TCP_conn=5
> 2024-08-05 19:22:03 No response received for 5 seconds
> 2024-08-05 19:22:03 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=1
> 2024-08-05 19:22:04 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=2
> 2024-08-05 19:22:05 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=3
> 2024-08-05 19:22:06 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=4
> 2024-08-05 19:22:07 Reply from www.google.com (142.250.195.196) on port 80 TCP_conn=5
> ```

Hi @SYSHIL,

Sorry, I had to be more concise. `tcping`'s printer comes in three forms:

1. plane printer (the one you have been working on)
2. json printer which is activated by the `-j` and `-j --pretty` flags
3. database printer which is activated by `--db` flag

When making changes to the way we display information, we need to ensure all of these printers outputting the same data to prevent any unexpected behavior for our users.

Now, what needs to be done is to implement the same for `json` and `db` printers.

Hopefully, you still have patience for this. 😄

**Comment by SYSHIL** [comment]
Hey @pouriyajamshidi, Yeah the reason I did not touch our json printer and database printer was because they already seem to be printing our current timestamp. [here](https://github.com/pouriyajamshidi/tcping/blob/master/statsprinter.go#L217-L218) and [here](https://github.com/pouriyajamshidi/tcping/blob/master/db.go#L184) so this -D flag would only extend the capability of the plane printer. Sorry if I'm misunderstanding this and thank you for bearing with me 🙏

**Comment by pouriyajamshidi** [comment]
> Hey @pouriyajamshidi, Yeah the reason I did not touch our json printer and database printer was because they already seem to be printing our current timestamp. [here](https://github.com/pouriyajamshidi/tcping/blob/master/statsprinter.go#L217-L218) and [here](https://github.com/pouriyajamshidi/tcping/blob/master/db.go#L184) so this -D flag would only extend the capability of the plane printer. Sorry if I'm misunderstanding this and thank you for bearing with me 🙏

ah, sorry

I will have a closer look tomorrow.

File: statsprinter.go [hunk 2]
```diff
@@ -26,10 +36,54 @@ var (
...
+}
 
 func (p *planePrinter) printStart(hostname string, port uint16) {
-	colorLightCyan("TCPinging %s on port %d\n", hostname, port)
+	options := &PrintOptions{
+		Color:         LightCyan,
+		Message:       fmt.Sprintf("TCPinging %s on port %d\n", hostname, port),
+		ShowTimestamp: p.ShowTimestamp,
+	}
+	printReply(options)
```
Comment (pouriyajamshidi) [inline statsprinter.go:86]: printStart does not need a timestamp

File: statsprinter.go [hunk 3]
```diff
@@ -26,10 +36,54 @@ var (
...
-type planePrinter struct{}
+type planePrinter struct {
+	ShowTimestamp *bool
+}
+
+type PrintOptions struct {
+	Color         Color
+	Message       string
+	ShowTimestamp *bool
+}
```
Comment (pouriyajamshidi) [inline statsprinter.go:47]: why do you need to export this struct and its fields?

File: statsprinter.go [hunk 4]
```diff
@@ -159,44 +213,83 @@ func (p *planePrinter) printStatistics(t tcping) {
...
-		return
+		message = fmt.Sprintf("Reply from %s on port %d TCP_conn=%d\n", ip, port, streak)
+	} else {
+		message = fmt.Sprintf("Reply from %s (%s) on port %d TCP_conn=%d\n", hostname, ip, port, streak)
 	}
-
-	colorLightGreen("Reply from %s (%s) on port %d TCP_conn=%d time=%.3f ms\n",
-		hostname, ip, port, streak, rtt)
+	options := &PrintOptions{
+		Color:         Green,
```
Comment (pouriyajamshidi) [inline statsprinter.go:223]: this should be light green

File: statsprinter.go [hunk 5]
```diff
@@ -159,44 +213,83 @@ func (p *planePrinter) printStatistics(t tcping) {
...
 
 func (p *planePrinter) printProbeSuccess(hostname, ip string, port uint16, streak uint, rtt float32) {
+	var message string
 	if hostname == "" {
-		colorLightGreen("Reply from %s on port %d TCP_conn=%d time=%.3f ms\n",
-			ip, port, streak, rtt)
-		return
+		message = fmt.Sprintf("Reply from %s on port %d TCP_conn=%d\n", ip, port, streak)
+	} else {
+		message = fmt.Sprintf("Reply from %s (%s) on port %d TCP_conn=%d\n", hostname, ip, port, streak)
```
Comment (pouriyajamshidi) [inline statsprinter.go:220]: Why did you remove the RTT?

File: statsprinter.go [hunk 6]
```diff
@@ -159,44 +213,83 @@ func (p *planePrinter) printStatistics(t tcping) {
...
 
 func (p *planePrinter) printVersion() {
-	colorGreen("TCPING version %s\n", version)
+	message := fmt.Sprintf("TCPING version %s\n", version)
+	options := &PrintOptions{
+		Color:         LightYellow,
+		Message:       message,
+		ShowTimestamp: p.ShowTimestamp,
+	}
+	printReply(options)
```
Comment (pouriyajamshidi) [inline statsprinter.go:292]: This does not need a timestamp

File: statsprinter.go [hunk 7]
```diff
@@ -159,44 +213,83 @@ func (p *planePrinter) printStatistics(t tcping) {
...
 
 func (p *planePrinter) printInfo(format string, args ...any) {
-	colorLightBlue(format+"\n", args...)
+	message := fmt.Sprintf(format+"\n", args...)
+	options := &PrintOptions{
+		Color:         LightCyan,
+		Message:       message,
+		ShowTimestamp: p.ShowTimestamp,
+	}
+	printReply(options)
```
Comment (pouriyajamshidi) [inline statsprinter.go:272]: This does not need a timestamp

File: statsprinter.go [hunk 8]
```diff
@@ -159,44 +213,83 @@ func (p *planePrinter) printStatistics(t tcping) {
...
 
 func (p *planePrinter) printError(format string, args ...any) {
-	colorRed(format+"\n", args...)
+	message := fmt.Sprintf(format+"\n", args...)
+	options := &PrintOptions{
+		Color:         Red,
+		Message:       message,
+		ShowTimestamp: p.ShowTimestamp,
+	}
+	printReply(options)
```
Comment (pouriyajamshidi) [inline statsprinter.go:282]: This does not need a timestamp

File: statsprinter.go [hunk 9]
```diff
@@ -159,44 +213,83 @@ func (p *planePrinter) printStatistics(t tcping) {
...
 
 func (p *planePrinter) printTotalDownTime(downtime time.Duration) {
-	colorYellow("No response received for %s\n", durationToString(downtime))
+	message := fmt.Sprintf("No response received for %s\n", durationToString(downtime))
+	options := &PrintOptions{
+		Color:         Yellow,
+		Message:       message,
+		ShowTimestamp: p.ShowTimestamp,
+	}
+	printReply(options)
```
Comment (pouriyajamshidi) [inline statsprinter.go:252]: This does not need a timestamp

**Review by pouriyajamshidi — CHANGES_REQUESTED** [review]

File: statsprinter.go (hunk 2 above)
Comment (SYSHIL) [reply statsprinter.go:86]: done, thanks

File: statsprinter.go (hunk 5 above)
Comment (SYSHIL) [reply statsprinter.go:220]: i'm sorry i must have missed that, fixing now

File: statsprinter.go (hunk 9 above)
Comment (SYSHIL) [reply statsprinter.go:252]: done

File: statsprinter.go (hunk 6 above)
Comment (SYSHIL) [reply statsprinter.go:292]: done thanks

File: statsprinter.go (hunk 8 above)
Comment (SYSHIL) [reply statsprinter.go:282]: could u please elaborate? 
We print the timestamp for succesful probes but not for unsuccesful probes?

File: statsprinter.go (hunk 3 above)
Comment (SYSHIL) [reply statsprinter.go:47]: We don't need to, This is my mistake...I've fixed it.

File: statsprinter.go (hunk 7 above)
Comment (SYSHIL) [reply statsprinter.go]: fixed

File: statsprinter.go (hunk 8 above)
Comment (pouriyajamshidi) [reply statsprinter.go]: errors regarding probe failures use `printProbeFail`. The `printError` is used for reporting errors, usually before exiting. You can see the examples in tcping.go

Comment (SYSHIL) [reply statsprinter.go]: ahh got it, thank u

**Comment by SYSHIL** [comment]
Hey @pouriyajamshidi could u please check now, Thank you :)

**Comment by pouriyajamshidi** [comment]
> Hey @pouriyajamshidi could u please check now, Thank you :)

Sure. Will get to this in a day or two max.

**Comment by pouriyajamshidi** [comment]
Hey @SYSHIL,

I can see the reviews I left for not needing to timestamp certain messages are marked as solved but they are still using the `PrintOptions` struct with `showTimestamp` set to false.

I do not see any benefit in doing that. Please just revert them to what they were to keep the code simple.

**Comment by SYSHIL** [comment]
Ahh sorry @pouriyajamshidi, somehow I missed that. Removed it now

File: statsprinter.go [hunk 10]
```diff
@@ -159,44 +212,80 @@ func (p *planePrinter) printStatistics(t tcping) {
...
 }
 
 func (p *planePrinter) printVersion() {
-	colorGreen("TCPING version %s\n", version)
+	message := fmt.Sprintf("TCPING version %s\n", version)
+	options := &PrintOptions{
+		color:   LightYellow,
+		message: message,
+	}
+	printReply(options)
```
Comment (pouriyajamshidi) [inline statsprinter.go:288]: these are still here? Similar to the rest

**Review by pouriyajamshidi — CHANGES_REQUESTED** [review]

**Comment by pouriyajamshidi** [comment]
Hey @SYSHIL

Could you please review your PR once while you are in a good head-space?

I think you have missed a couple of changes I have requested and every time I see "done or fixed", I take a look at the code and it confuses me more. 😄 

Thanks

File: statsprinter.go (hunk 10 above)
Comment (SYSHIL) [reply statsprinter.go]: I think you meant that it was carrying the wrong color?
Yeah I fixed that.

**Comment by SYSHIL** [comment]
Sure @pouriyajamshidi I will review the entire PR and ping you, Sorry for bothering you with this. I will make sure this doesn't happen again, Just new to the tool and its codebase 🙏

**Comment by SYSHIL** [comment]
Alright time in output lines only shows up upon `printprobefailure` and `printprobesuccess` but I don't think I quite understood this comment.

> I can see the reviews I left for not needing to timestamp certain messages are marked as solved but they are still using the `PrintOptions` struct with `showTimestamp` set to false.

They do use `PrintOptions` stuct but the showTimestamp field is not set to false its set to nil by default because its a pointer to a boolean. I can see the need to not tweak the existing function at all and revert my changes but I thought it would be cleaner to just have one way of printing things through the plane printer in this case it is through the `printReply` function.

**Comment by pouriyajamshidi** [comment]
> Alright time in output lines only shows up upon `printprobefailure` and `printprobesuccess` but I don't think I quite understood this comment. Could you please elaborate on this @pouriyajamshidi ?
> 
> > I can see the reviews I left for not needing to timestamp certain messages are marked as solved but they are still using the `PrintOptions` struct with `showTimestamp` set to false.
> 
> They do use `PrintOptions` struct but the `showTimestamp` field is not set to `false` its set to `nil` by default because its a pointer to a boolean. I can see the need to not tweak the existing function at all and revert my changes but I thought it would be cleaner to just have one way of printing things through the plane printer in this case it is through the `printReply` function.

This is what I meant.

I don't see any point in using `PrintOptions` when you don't need to.

**Comment by SYSHIL** [comment]
> This is what I meant.
> I don't see any point in using PrintOptions when you don't need it.

Okay @pouriyajamshidi I'm sorry about the confusion my initial idea was to just have one way of printing things through the planPrinter and that's why I added printOptions in case there's a need to extend it but I also agree with keeping the code simple, I've reverted my changes in functions where printoptions isn't really helping us

File: statsprinter.go [hunk 11]
```diff
@@ -26,7 +36,46 @@ var (
...
+	message       string
+	showTimestamp *bool
+}
+
+func printReply(options *PrintOptions) {
+	if options.showTimestamp != nil && *options.showTimestamp {
+		t := time.Now().Format(timeFormat)
+		options.message = fmt.Sprintf("%s %s", t, options.message)
+	}
+	switch options.color {
```
Comment (SYSHIL) [inline statsprinter.go:54]: Only 2 functions use printReply now, So we should be okay to remove most of these switchcases and leave something like Yellow to be the default color. What do you think @pouriyajamshidi ?

Comment (pouriyajamshidi) [reply statsprinter.go:54]: hmmmm, right.

Or you could also just pass the color as an argument to it instead of using enums

File: tcping.go [hunk 12]
```diff
@@ -227,7 +227,7 @@ func usage() {
 }
 
 // setPrinter selects the printer
-func setPrinter(tcping *tcping, outputJSON, prettyJSON *bool, outputDb *string, args []string) {
+func setPrinter(tcping *tcping, outputJSON, prettyJSON *bool, timeStamp *bool, outputDb *string, args []string) {
```
Comment (pouriyajamshidi) [inline tcping.go:230]: by the way, this is why CI is complaining about complex code. Although, I do not expect you to solve this.

File: statsprinter.go (hunk 11 above)
Comment (pouriyajamshidi) [reply statsprinter.go:54]: You are not wrong. But we are looking at the problem from two different perspectives.

What I mean to say is, we do not need all of this "fancy" stuff to implement this feature.

If we only add an extra check for the `-D` flag in `printProbeSuccess` and `printProbeFail`, we then do not need enums, `printReply` function and the `PrintOptions` struct. 

A quick and dirty (yet working) implementation for `printProbeSuccess` is:

```go
func (p *planePrinter) printProbeSuccess(hostname, ip string, port uint16, streak uint, rtt float32) {
	if hostname == "" && *p.showTimestamp == false {
		colorLightGreen("Reply from %s on port %d TCP_conn=%d time=%.3f ms\n",
			ip, port, streak, rtt)
	} else if hostname == "" && *p.showTimestamp {
		t := time.Now().Format(timeFormat)
		colorLightGreen("%s Reply from %s on port %d TCP_conn=%d time=%.3f ms\n",
			t, ip, port, streak, rtt)
	} else if hostname != "" && *p.showTimestamp == false {
		colorLightGreen("Reply from %s (%s) on port %d TCP_conn=%d time=%.3f ms\n",
			hostname, ip, port, streak, rtt)
	} else {
		t := time.Now().Format(timeFormat)
		colorLightGreen("%s Reply from %s (%s) on port %d TCP_conn=%d time=%.3f ms\n",
			t, hostname, ip, port, streak, rtt)
	}
}
```

**Review by pouriyajamshidi — CHANGES_REQUESTED** [review]

File: tcping.go (hunk 12 above)
Comment (SYSHIL) [reply tcping.go:230]: oh okk thanks for the info

File: statsprinter.go (hunk 11 above)
Comment (SYSHIL) [reply statsprinter.go]: okay I understand, I've tweaked my code to follow a pattern similar to this

File: statsprinter.go [hunk 13]
```diff
@@ -10,9 +10,19 @@ import (
...
+	timeFormat        = "2006-01-02 15:04:05"
+	hourFormat        = "15:04:05"
+	Yellow      Color = "yellow"
+	Cyan        Color = "cyan"
+	Green       Color = "green"
+	Red         Color = "red"
+	LightYellow Color = "light_yellow"
+	LightGreen  Color = "light_green"
+	LightCyan   Color = "light_cyan"
+	LightBlue   Color = "light_blue"
```
[the remark this answers is not in the document]
Comment (SYSHIL) [reply statsprinter.go]: this is removed along with the enums

File: statsprinter.go [hunk 14]
```diff
@@ -159,24 +165,43 @@ func (p *planePrinter) printStatistics(t tcping) {
 }
 
 func (p *planePrinter) printProbeSuccess(hostname, ip string, port uint16, streak uint, rtt float32) {
+	t := ""
```
Comment (pouriyajamshidi) [inline statsprinter.go:168]:
```suggestion
	timestamp := ""
```

Comment (pouriyajamshidi) [reply statsprinter.go:168]: Let's make this more readable

File: statsprinter.go [hunk 15]
```diff
@@ -159,24 +165,43 @@ func (p *planePrinter) printStatistics(t tcping) {
...
+			colorLightGreen("%s Reply from %s (%s) on port %d TCP_conn=%d time=%.3f ms\n", t, hostname, ip, port, streak, rtt)
+		}
 	}
-
-	colorLightGreen("Reply from %s (%s) on port %d TCP_conn=%d time=%.3f ms\n",
-		hostname, ip, port, streak, rtt)
 }
 
 func (p *planePrinter) printProbeFail(hostname, ip string, port uint16, streak uint) {
+	t := ""
```
[the remark this answers is not in the document]
Comment (pouriyajamshidi) [reply statsprinter.go:188]: same here

**Comment by SYSHIL** [comment]
I've made the appropriate changes @pouriyajamshidi, Sorry about the initial confusion with respect to requirements....Would love to collaborate soon if there any new issues I could pick up :)

**Comment by pouriyajamshidi** [comment]
> I've made the appropriate changes @pouriyajamshidi, Sorry about the initial confusion with respect to requirements....Would love to collaborate soon if there any new issues I could pick up :)

Thanks for pushing this through to the end.

Perhaps solving the CI error next would be a good move? 

Nonetheless, appreciate the help

**Review by pouriyajamshidi — APPROVED** [review]
LGTM :shipit: 

closes #215
